### ISSUE TITLE
Revisit dynamic collection initializers in nullable

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#33441]
https://github.com/dotnet/roslyn/blob/0530f607e9f2ef1571b82f2c1be07893bed02a72/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L5363

This statement isn't true, dynamic collection initializers always hit this function. We need to look at this and re-evaluate whether these need to be handled differently, and some tests should be added.

**Comment by 333fred** [comment]
```C#
using System.Collections.Generic;
#nullable enable
public class C
{
    public static void M(dynamic d, object? o)
    {
        _ = new List<int>() { d(o.ToString()) };
    }
}
```
We should be warning on the `o.ToString()` call but we are not.

**Comment by 333fred** [comment]
```C#
#nullable enable
public class C
{
    public dynamic D { get; set; } = null!;
    public static void M(object? o)
    {
        _ = new C() { D = { X = { Y = o.ToString() } } };
    }
}
```

We should be warning on the `o.ToString()` call. This bug is because `VisitObjectElementInitializer` isn't handling when `X` in `X = { Y = o.ToString() }` is a `BoundDynamicObjectInitializerMember`.
